src/util/logging.cpp

```diff
@@ -43,41 +46,114 @@ Q_DECLARE_FLAGS(WriteFlags, WriteFlag)
+        levelName = QStringLiteral("Warning");
+        break;
+    case QtCriticalMsg:
+        levelName = QStringLiteral("Critical");
+        break;
+    case QtFatalMsg:
+        levelName = QStringLiteral("Fatal");
+        break;
+    }
+
+    return levelName + QStringLiteral(" [") + threadName + QStringLiteral("] ") + message;
+}
+
+/// Actually write a log message to a file.
+inline void writeToFile(
+        QtMsgType type,
+        const QMessageLogContext& context,
+        const QString& message,
+        const QString& threadName,
+        bool flush) {
+    QString formattedMessageStr =
+            formatLogFileMessage(type, context, message, threadName) +
+            QChar('\n');
+    QByteArray formattedMessage = formattedMessageStr.toLocal8Bit();
+
+    QMutexLocker locked(&s_mutexLogfile);
+    // Writing to a closed QFile could cause an infinite recursive loop
+    // by logging to qWarning!
+    if (s_logfile.isOpen()) {
+        const int written = s_logfile.write(formattedMessage);
+        Q_UNUSED(written);
+        DEBUG_ASSERT(written == formattedMessage.size());
+        if (flush) {
+            const bool flushed = s_logfile.flush();
+            Q_UNUSED(flushed);
+            DEBUG_ASSERT(flushed);
+        }
+    }
+}
+
+/// Actually write a log message to stderr.
+inline void writeToStdErr(
+        QtMsgType type,
+        const QMessageLogContext& context,
+        const QString& message,
+        const QString& threadName,
+        bool flush) {
+    QString formattedMessageStr = qFormatLogMessage(type, context, message) + QChar('\n');
+    const QByteArray formattedMessage =
+            formattedMessageStr.replace(kThreadNamePattern, threadName)
+                    .toLocal8Bit();
+
+    QMutexLocker locked(&s_mutexStdErr);
+    const size_t written = fwrite(
+            formattedMessage.constData(), sizeof(char), formattedMessage.size(), stderr);
+    Q_UNUSED(written);
+    DEBUG_ASSERT(written == static_cast<size_t>(formattedMessage.size()));
+    if (flush) {
+        // Flushing stderr might not be necessary, because message
+        // should end with a newline character. Flushing occurs
+        // only infrequently (log level >= Critical), so better safe
+        // than sorry.
+        const int ret = fflush(stderr);
+        Q_UNUSED(ret);
+        DEBUG_ASSERT(ret == 0);
+    }
+}
+
+/// Handles writing to stderr and the log file.
 inline void writeToLog(
-        const QByteArray& message,
+        QtMsgType type,
+        const QMessageLogContext& context,
+        const QString& message,
         WriteFlags flags) {
     DEBUG_ASSERT(!message.isEmpty());
     DEBUG_ASSERT(flags & (WriteFlag::StdErr | WriteFlag::File));
+
+    QString threadName = QThread::currentThread()->objectName();
+    if (threadName.isEmpty()) {
+        threadName = QStringLiteral("0x") +
+                QString::number(
+                        reinterpret_cast<size_t>(QThread::currentThread()), 16);
```

Comment:
QTextStream (constructed with a QString pointer for the underlying buffer) should work:
https://doc.qt.io/qt-5/qtextstream.html#operator-lt-lt-18